### ISSUE TITLE
Client cannot authenticate via:[TOKEN, KERBEROS]

### ISSUE THREAD

**Issue by jasonTuZq** [linkedin/TonY#292]
hi, i'am trying the mnist-tensorflow example on my cluster (which has been kerberosized), the node which submit this tony-job shows log below:
```
...
19/05/17 19:31:25 INFO client.AHSProxy: Connecting to Application History server at ...
19/05/17 19:31:25 INFO client.RequestHedgingRMFailoverProxyProvider: Looking for the active RM in [rm1, rm2]...
19/05/17 19:31:25 INFO client.RequestHedgingRMFailoverProxyProvider: Found active RM [rm1]
19/05/17 19:31:41 INFO tony.TonyClient: Running with secure cluster mode. Fetching delegation tokens..
19/05/17 19:31:41 INFO tony.TonyClient: Fetching RM delegation token..
19/05/17 19:31:41 INFO tony.TonyClient: RM delegation token fetched.
19/05/17 19:31:41 INFO tony.TonyClient: Fetching HDFS delegation token for default namenode: hdfs://wcluster
19/05/17 19:31:41 INFO hdfs.DFSClient: Created HDFS_DELEGATION_TOKEN token xxx for public_dataetl on ha-hdfs:wcluster
19/05/17 19:31:41 INFO tony.TonyClient: Default HDFS delegation token fetched.
19/05/17 19:31:41 INFO hdfs.DFSClient: Created HDFS_DELEGATION_TOKEN token xxx for public_dataetl on ha-hdfs:wcluster
19/05/17 19:31:41 INFO tony.TonyClient: Fetched delegation token for history filesystem HDFS.
19/05/17 19:31:41 INFO tony.TonyClient: Successfully fetched tokens.
19/05/17 19:31:41 INFO tony.TonyClient: Completed setting up Application Master command {{JAVA_HOME}}/bin/java -Xmx1638m -Dyarn.app.container.log.dir=<LOG_DIR> com.linkedin.tony.ApplicationMaster 1><LOG_DIR>/amstdout.log 2><LOG_DIR>/amstderr.log
19/05/17 19:31:41 INFO tony.TonyClient: Submitting YARN application
19/05/17 19:31:42 INFO impl.TimelineClientImpl: Timeline service address: http://...
19/05/17 19:31:42 INFO impl.YarnClientImpl: Submitted application application_1557706034708_17871
19/05/17 19:31:42 INFO tony.TonyClient: URL to track running application (will proxy to TensorBoard once it has started): http://...
19/05/17 19:31:42 INFO tony.TonyClient: ResourceManager web address for application: http://...
19/05/17 19:31:59 INFO tony.TonyClient: AM host: ...
19/05/17 19:31:59 INFO tony.TonyClient: AM RPC port: 12043
19/05/17 19:32:01 INFO tony.TonyClient: Tasks Status Updated: [TaskInfo] name: worker index: 0 url: ... status: RUNNING
19/05/17 19:32:01 INFO tony.TonyClient: Tasks Status Updated: [TaskInfo] name: ps index: 0 url: ... status: RUNNING
19/05/17 19:32:01 INFO tony.TonyClient: Logs for ps 0 at: http://...
19/05/17 19:32:01 INFO tony.TonyClient: Logs for worker 0 at: http://...
```
this seems allright, while the logs in the ps & worker node both show `java.io.IOException: org.apache.hadoop.security.AccessControlException: Client cannot authenticate via:[TOKEN, KERBEROS]` , more is below:
```
2019-05-17 19:35:09 WARN  Client:715 - Exception encountered while connecting to the server : org.apache.hadoop.security.AccessControlException: Client cannot authenticate via:[TOKEN, KERBEROS]
2019-05-17 19:35:09 ERROR TaskMonitor:49 - Encountered exception updating metrics
java.lang.reflect.UndeclaredThrowableException
	at com.sun.proxy.$Proxy7.updateMetrics(Unknown Source)
	at com.linkedin.tony.TaskMonitor.run(TaskMonitor.java:47)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: Failed on local exception: java.io.IOException: org.apache.hadoop.security.AccessControlException: Client cannot authenticate via:[TOKEN, KERBEROS]; Host Details : local host is: "..."; destination host is: "...":45607; 
	at org.apache.hadoop.net.NetUtils.wrapException(NetUtils.java:785)
	at org.apache.hadoop.ipc.Client.getRpcResponse(Client.java:1558)
	at org.apache.hadoop.ipc.Client.call(Client.java:1498)
	at org.apache.hadoop.ipc.Client.call(Client.java:1398)
	at org.apache.hadoop.ipc.WritableRpcEngine$Invoker.invoke(WritableRpcEngine.java:242)
```
this seems having problem to authenticate via:[TOKEN, KERBEROS], any way to fix would be very appreciate.

**Comment by oliverhu** [comment]
@jasonTuZq is that failing your job?  I think this is related to the TonyMetrics for metrics data collection and should not be in the critical path.

@erwa ^^

**Comment by jasonTuZq** [comment]
> @jasonTuZq is that failing your job? I think this is related to the TonyMetrics for metrics data collection and should not be in the critical path.
> 
> @erwa ^^

@oliverhu this doesn't fail the job but the training doesn't seem to run,  the job doesn't get finished even after several hours & the print infos i add to the training py script don't show on the log of any ps or worker node.

**Comment by jasonTuZq** [comment]
below are more log that i forgot to paste, this seems to be a critical problem
```
WARN  Utils:129 - Function didn't return non-null within 0 seconds.
ERROR TaskExecutor:119 - Failed to register worker with AM.
WARN  Client:715 - Exception encountered while connecting to the server : org.apache.hadoop.security.AccessControlException: Client cannot authenticate via:[TOKEN, KERBEROS]
ERROR TaskExecutor:277 - [worker:1] Failed to send Heart Beat.
java.io.IOException: Failed on local exception: java.io.IOException: org.apache.hadoop.security.AccessControlException: Client cannot authenticate via:[TOKEN, KERBEROS]; Host Details : local host is: "..."; destination host is: "...":10346; 
	at org.apache.hadoop.net.NetUtils.wrapException(NetUtils.java:785)
```

**Comment by erwa** [comment]
This may be related to #284, which another user reported.

**Comment by erwa** [comment]
@jasonTuZq , is `hadoop.security.token.service.use_ip=true` in your cluster? (If it's not set, the default is `true`.)

**Comment by oliverhu** [comment]
Should be fixed in master HEAD.

**Comment by oliverhu** [comment]
I'm a bit surprised that we don't see this issue in our cluster tho 🤔  @erwa

**Comment by erwa** [comment]
@oliverhu, we probably don't see this issue because in our clusters, we set `hadoop.security.token.service.use_ip=false`.

**Comment by oliverhu** [comment]
Oh I see. I thought we use default in our cluster. Any idea why we set it to false manually? @erwa

**Comment by oliverhu** [comment]
Merged and published in 0.3.11

**Comment by erwa** [comment]
I don't know the context behind why we set `hadoop.security.token.service.use_ip=false` in our clusters. I suggest checking with our SREs.
